smtpd/lka.c

```diff
@@ -435,8 +435,7 @@ lka(void)
 
 	config_process(PROC_LKA);
 
-	if (initgroups(pw->pw_name, pw->pw_gid) ||
```

Comment:
Hi @catern, heads up I'm not a developer of OpenSMTPD, I only have a question about the best practice approach here.

My hypothesis is that using `initgroups` to maintain _only the intended groups_ of the _target user_ seems correct and this line should not be removed. Using `initgroups` maintains the expectation that a user can be placed into a group used to restrict privileges.

According to this hypothesis the use of `setgroups` in `./smtpd/control.c` should be changed to `initgroups` (vs. being removed).

Curious about your opinion/guidance, thanks!

Reply:
initgroups resets the current supplementary groups back to "default". That's a problem because it necessitates that we start with the privileges required to change groups, which is root-equivalent. (I guess there could be a Unix where you can call something like initgroups without privs, but that's not the case on Linux or BSD)

But, ultimately, resetting groups shouldn't be necessary anyway: we are already assuming with these changes that the process is started with the correct user and primary group, so the process should likewise be started with the correct supplementary groups, instead of changing after it starts running.

This all comes from the philosophy that we should start already sandboxed and without privileges, instead of dropping privs after we start. That doesn't fit upstream OpenSMTPD, but I think it's correct.